tools/circt-lec/circt-lec.cpp

```diff
@@ -0,0 +1,163 @@
+/// orchestrates the work to be done. It first parses the input files, then it
+/// traverses their IR to export the logical constraints from the given circuit
+/// description to an internal circuit representation, lastly, these will be
+/// compared and solved for equivalence.
+static mlir::LogicalResult executeLEC(mlir::MLIRContext &context) {
+  // Parse the provided input files.
+  if (verbose)
+    lec::outs() << "Parsing input file\n";
+  mlir::OwningOpRef<mlir::ModuleOp> file1 =
+      mlir::parseSourceFile<mlir::ModuleOp>(fileName1, &context);
+  if (!file1)
+    return mlir::failure();
+
+  mlir::OwningOpRef<mlir::ModuleOp> file2;
+  if (!fileName2.empty()) {
+    if (verbose)
+      lec::outs() << "Parsing second input file\n";
+    file2 = mlir::parseSourceFile<mlir::ModuleOp>(fileName2, &context);
+    if (!file2)
+      return mlir::failure();
+  } else if (verbose)
+    lec::outs() << "Second input file not specified\n";
+
+  // Initiliaze the constraints solver and the circuits to be compared.
+  Solver s(&context, statisticsOpt);
+  Solver::Circuit *c1 = s.addCircuit(moduleName1);
+  Solver::Circuit *c2 = s.addCircuit(moduleName2);
+
+  // Initialize a logic exporter for the first circuit then run it on the
+  // top-level module of the first input file.
+  if (verbose)
+    lec::outs() << "Analyzing the first circuit\n";
+  auto exporter = std::make_unique<LogicExporter>(moduleName1, c1);
+  mlir::ModuleOp m = file1.get();
+  if (failed(exporter->run(m)))
+    return mlir::failure();
+
+  // Repeat the same procedure for the second circuit.
+  if (verbose)
+    lec::outs() << "Analyzing the second circuit\n";
+  auto exporter2 = std::make_unique<LogicExporter>(moduleName2, c2);
+  // In case a second input file was not specified, the first input file will
+  // be used instead.
+  mlir::ModuleOp m2 = fileName2.empty() ? m : file2.get();
+  if (failed(exporter2->run(m2)))
+    return mlir::failure();
+
+  // The logical constraints have been exported to their respective circuit
+  // representations and can now be solved for equivalence.
+  if (verbose)
+    lec::outs() << "Solving constraints\n";
+  return s.solve();
+}
+
+/// The entry point for the `circt-lec` tool:
+/// configures and parses the command-line options,
+/// registers all dialects within a MLIR context,
+/// and calls the `executeLEC` function to do the actual work.
+int main(int argc, char **argv) {
+  // Configure the relevant command-line options.
+  cl::HideUnrelatedOptions(mainCategory);
+  mlir::registerMLIRContextCLOptions();
+  cl::AddExtraVersionPrinter(
+      [](llvm::raw_ostream &os) { os << circt::getCirctVersion() << '\n'; });
+
+  // Parse the command-line options provided by the user.
+  cl::ParseCommandLineOptions(
+      argc, argv,
+      "circt-lec - logical equivalence checker\n\n"
+      "\tThis tool compares two input circuit descriptions to determine whether"
+      " they are logically equivalent.\n");
```

Comment:
Very cool. I love that CIRCT has gotten to a point where we can have a tool print this help output :sunglasses: